app_dart/lib/src/request_handlers/task_status_subscription.dart

```diff
@@ -0,0 +1,127 @@
+    }
+    final String? newStatus = buildPushMessage.build?.status as String?;
+    if (newStatus == null) {
+      throw const BadRequestException('userData does not contain builder_name');
+    }
+
+    final DatastoreService datastore = datastoreProvider(config.db);
+
+    if (newStatus != Task.statusSucceeded && newStatus != Task.statusFailed) {
+      throw const BadRequestException('NewStatus can be one of "Succeeded", "Failed"');
+    }
+
+    final Task task = await _getTaskFromNamedParams(datastore, builderName, commitBranch, commitSha);
+
+    task.status = newStatus;
+    task.endTimestamp = DateTime.now().millisecondsSinceEpoch;
+
+    await datastore.insert(<Task>[task]);
+    return Body.empty;
+  }
+
+  /// Retrieve [Task] from [DatastoreService] when given [commitSha], [commitBranch], and [builderName].
+  ///
+  /// This is used when the DeviceLab test runner is uploading results to Cocoon for runs on LUCI.
+  /// LUCI does not know the [Key] assigned to task when scheduling the build, but Cocoon can
+  /// lookup the task based on these key values.
+  ///
+  /// To lookup the value, we construct the ancestor key, which corresponds to the [Commit].
+  /// Then we query the tasks with that ancestor key and search for the one that matches the builder name.
+  Future<Task> _getTaskFromNamedParams(
+      DatastoreService datastore, String builderName, String gitBranch, String gitSha) async {
+    final Key<String> commitKey = await _constructCommitKey(datastore, gitBranch, gitSha);
+
+    final Query<Task> query = datastore.db.query<Task>(ancestorKey: commitKey);
+    final List<Task> initialTasks = await query.run().toList();
+    log.fine('Found ${initialTasks.length} tasks for commit');
+    final List<Task> tasks = <Task>[];
+    log.fine('Searching for task with builderName=$builderName');
+    for (Task task in initialTasks) {
+      if (task.builderName == builderName || task.name == builderName) {
+        tasks.add(task);
+      }
+    }
+
+    if (tasks.length != 1) {
+      log.severe('Found ${tasks.length} entries for builder $builderName');
+      throw InternalServerError('Expected to find 1 task for $builderName, but found ${tasks.length}');
+    }
+
+    return tasks.first;
+  }
+
+  /// Construct the Datastore key for [Commit] that is the ancestor to this [Task].
+  ///
+  /// Throws [BadRequestException] if the given git branch does not exist in [CocoonConfig].
+  Future<Key<String>> _constructCommitKey(DatastoreService datastore, String commitBranch, String commitSha) async {
+    final List<String> flutterBranches = await config.flutterBranches;
+    if (!flutterBranches.contains(commitBranch)) {
+      throw BadRequestException('Failed to find flutter/flutter branch: $commitBranch\n'
+          'If this is a valid branch, '
+          'see https://github.com/flutter/cocoon/tree/master/app_dart#branching-support-for-flutter-repo');
```

Comment:
```suggestion
          'see https://github.com/flutter/cocoon/tree/main/app_dart#branching-support-for-flutter-repo');
```

Reply:
Are these docs still up to date? I don't think we want to check config.flutterBranches, and instead rely on the key lookup

Reply:
Removed the `config.flutterBrances` use.